File: examples/OrderService.java
package com.example.orders;

import java.math.BigDecimal;

@interface Transactional {
    boolean readOnly();
}

class InvalidPromoException extends Exception {
    InvalidPromoException(String message) {
        super(message);
    }
}

class Promo {
    BigDecimal getAmount() {
        return new BigDecimal("5.00");
    }
}

class Order {
    BigDecimal getTotal() {
        return new BigDecimal("100.00");
    }
}

class PromoRepository {
    Promo findByCode(String code) {
        return new Promo();
    }
}

abstract class BaseService {
    abstract BigDecimal applyDiscount(Order order, String code) throws InvalidPromoException;

    protected void validate(Promo promo) throws InvalidPromoException {
        if (promo == null) {
            throw new InvalidPromoException("unknown promo");
        }
    }
}

public class OrderService extends BaseService {

    private final PromoRepository promoRepository = new PromoRepository();

    /**
     * Applies a promotional discount to an order.
     *
     * @param order the order to discount
     * @param code the promo code to apply
     * @return the new total after discount
     * @throws InvalidPromoException if the code is expired
     */
    @Override
    @Transactional(readOnly = false)
    public BigDecimal applyDiscount(Order order, String code) throws InvalidPromoException {
        Promo promo = promoRepository.findByCode(code);
        validate(promo);
        return order.getTotal().subtract(promo.getAmount());
    }
}

File: examples/OrderService.cs
using System;
using System.Collections.Generic;

namespace Example.Orders
{
    public class TransactionalAttribute : Attribute
    {
        public bool ReadOnly { get; set; }
    }

    public class Promo
    {
        public decimal GetAmount() => 5.00m;
    }

    public class Order
    {
        public decimal GetTotal() => 100.00m;
    }

    public class PromoRepository
    {
        public Promo FindByCode(string code) => new Promo();
    }

    public abstract class BaseService
    {
        public abstract decimal ApplyDiscount(Order order, string code);

        protected void Validate(Promo promo)
        {
            if (promo == null) throw new ArgumentNullException(nameof(promo));
        }
    }

    public class OrderService : BaseService
    {
        private readonly PromoRepository _promoRepository = new PromoRepository();

        /// <summary>
        /// Applies a promotional discount to an order.
        /// </summary>
        /// <param name="order">the order to discount</param>
        /// <param name="code">the promo code to apply</param>
        /// <returns>the new total after discount</returns>
        [Transactional(ReadOnly = false)]
        public override decimal ApplyDiscount(Order order, string code)
        {
            Promo promo = _promoRepository.FindByCode(code);
            Validate(promo);
            return order.GetTotal() - promo.GetAmount();
        }
    }
}
